src/platform/cc13xx_26xx/FactoryDataProvider.cpp

```diff
@@ -0,0 +1,333 @@
+    memcpy(&vendorId, mFactoryData.vendor_id.data, mFactoryData.vendor_id.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetProductName(char * buf, size_t bufSize)
+{
+    ReturnErrorCodeIf(bufSize < mFactoryData.product_name.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.product_name.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memset(buf, 0, bufSize);
+    memcpy(buf, mFactoryData.product_name.data, mFactoryData.product_name.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetProductId(uint16_t & productId)
+{
+    ReturnErrorCodeIf(!mFactoryData.product_id.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(&productId, mFactoryData.product_id.data, mFactoryData.product_id.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetPartNumber(char * buf, size_t bufSize)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+CHIP_ERROR FactoryDataProvider::GetProductURL(char * buf, size_t bufSize)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+CHIP_ERROR FactoryDataProvider::GetProductLabel(char * buf, size_t bufSize)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR FactoryDataProvider::GetSerialNumber(char * buf, size_t bufSize)
+{
+    ReturnErrorCodeIf(bufSize < mFactoryData.serial_number.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.serial_number.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memset(buf, 0, bufSize);
+    memcpy(buf, mFactoryData.serial_number.data, mFactoryData.serial_number.len);
+    return CHIP_NO_ERROR;
+}
+
+// #pragma GCC push_options
+// #pragma GCC optimize("O0")
+CHIP_ERROR FactoryDataProvider::GetManufacturingDate(uint16_t & year, uint8_t & month, uint8_t & day)
+{
+    ReturnErrorCodeIf(!mFactoryData.manufacturing_date.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    uint8_t tmp[10] = { 0 };
+    memcpy(tmp, mFactoryData.manufacturing_date.data, 10);
+    year  = ((tmp[0] - 0x30) * 1000) + ((tmp[1] - 0x30) * 100) + ((tmp[2] - 0x30) * 10) + (tmp[3] - 0x30);
+    month = ((tmp[5] - 0x30) * 10) + (tmp[6] - 0x30);
+    day   = ((tmp[8] - 0x30) * 10) + (tmp[9] - 0x30);
+    return CHIP_NO_ERROR;
+}
+// #pragma GCC pop_options
+
+CHIP_ERROR FactoryDataProvider::GetHardwareVersion(uint16_t & hardwareVersion)
+{
+    ReturnErrorCodeIf(!mFactoryData.hw_ver.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(&hardwareVersion, mFactoryData.hw_ver.data, mFactoryData.hw_ver.len);
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR FactoryDataProvider::GetHardwareVersionString(char * buf, size_t bufSize)
+{
+    ReturnErrorCodeIf(bufSize < mFactoryData.hw_ver_str.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.hw_ver_str.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memset(buf, 0, bufSize);
+    memcpy(buf, mFactoryData.hw_ver_str.data, mFactoryData.hw_ver_str.len);
```

Comment:
Does `mFactoryData.hw_ver_str.len` include the null terminator?

Reply:
Does not include a null terminator, refer to comments on earlier elements

Reply:
See comments above.